language/stdlib/modules/Event.move

```diff
@@ -69,6 +75,189 @@ module Event {
     public fun destroy_handle<T: copyable>(handle: EventHandle<T>) {
         EventHandle<T> { counter: _, guid: _ } = handle;
     }
+
+    // ****************** SPECIFICATIONS *******************
+
+    /// # Module specifications
+    spec module {
+        /// Helper function that returns whether or not an EventHandleGenerator is
+        /// initilaized at the given address `addr`.
+        define initialized_ehg(addr: address): bool {
+            exists<EventHandleGenerator>(addr)
+        }
+        /// Helper function that returns the EventHandleGenerator at `addr`.
+        define ehg(addr: address): EventHandleGenerator {
+            global<EventHandleGenerator>(addr)
+        }
+        /// Helper function that returns the serialized counter of the `EventHandleGenerator` ehg.
+        define serialized_ehg_counter(ehg: EventHandleGenerator): vector<u8> {
+            LCS::serialize(ehg.counter)
+        }
+        /// Helper function that returns the serialized address of the `EventHandleGenerator` ehg.
+        define serialized_ehg_addr(ehg: EventHandleGenerator): vector<u8> {
+            LCS::serialize(ehg.addr)
+        }
+    }
+
+    /// ## Management of EventHandleGenerators
+
+    spec module {
+        /// Variable that is true whenever an `EventHandleGenerator` is destroyed.
+        /// This should never to true to preserve uniqueness of EventHandleGenerators.
+        global ehg_destroyed: bool;
+    }
+    spec struct EventHandleGenerator {
+        /// Updates the ehg_destroyed variable to true if an
+        /// `EventHandleGenerator` is ever unpacked.
+        invariant pack ehg_destroyed = ehg_destroyed;
+        invariant unpack ehg_destroyed = true;
+    }
+    spec schema EventHandleGeneratorNeverDestroyed {
+        /// **Informally:** No EventHandleGenerator should ever be destroyed.
+        invariant module !ehg_destroyed;
+    }
+    spec schema EventHandleGeneratorAtSameAddress {
+        /// **Informally:** An EventHandleGenerator should be located at `addr` and is never moved.
+        invariant module forall addr: address where initialized_ehg(addr): ehg(addr).addr == addr;
+    }
+    spec module {
```

Comment:
It seems obvious that fresh_guid cannot violate this property!  Must investigate...